Automated Fabric Test Config Generation

jpanasiukTT opened the pull request:
### Ticket
https://github.com/tenstorrent/tt-metal/issues/32258

### Problem description
- Mesh Graph Descriptors (MGDs) for multi-chip fabric topologies were manually written and maintained, making them error-prone and difficult to keep in sync with physical cabling configurations
- No automated way to generate MGDs from cabling descriptors, leading to potential inconsistencies between hardware setup and software configuration
- Lack of comprehensive testing to validate that generated MGDs correctly reflect the physical topology
- Manual MGD creation doesn't scale for complex topologies (e.g., 16-node ClosetBox with multiple superpods and inter-superpod connections)

### What's changed
Added generate_mgd tool:
- CLI tool that converts cabling descriptors to Mesh Graph Descriptors
- Automatically computes mesh instances, connections, and channel counts from physical topology
- Supports multiple node types (N300, P150, Galaxy, etc.)
- Provides file output and programmatic API

Added GTest validation suite:
- Tests MGD generation for 16-node ClosetBox topology and dual t3k, easily extendable to future topologies
- Validates mesh counts, architecture, and connection topology
- Verifies intra-superpod and inter-superpod connections

CMake and documentation:
- Added README with usage examples

### Checklist
- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] (For models and ops writers) [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes (if applicable) See [recommended dev flow](https://github.com/tenstorrent/tt-metal/blob/main/models/docs/MODEL_ADD.md#a-recommended-dev-flow-on-github-for-adding-new-models).
- [ ] [Galaxy quick](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-quick.yaml) CI passes (if applicable)
- [ ] [Galaxy demo tests, for Llama](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-demo-tests.yaml) CI passes, if applicable, because of current Llama work
- [ ] (For runtime and ops writers) [T3000 unit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-unit-tests.yaml) CI passes (if applicable, since this is run on push to main)
- [ ] (For models and ops writers) [T3000 demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-demo-tests.yaml) CI passes (if applicable, since this is required for release)
- [ ] New/Existing tests provide coverage for changes

tt-asaigal commented on tools/scaleout/generate_mgd/generate_mgd.cpp:
```diff
@@ -0,0 +1,488 @@
+    throw std::runtime_error("No node references found in graph template '" + template_name + "'");
+}
+
+CablingDescriptorInfo get_cabling_descriptor_info(const std::string_view cabling_descriptor_path) {
```
This function is not being used. Please remove from here and the header

tt-asaigal commented on tools/scaleout/generate_mgd/generate_mgd.cpp:
```diff
@@ -0,0 +1,488 @@
+
+    // Write cluster descriptor to a temporary file for CablingGenerator
+    // (CablingGenerator currently only accepts file paths)
+    std::string temp_file = "/tmp/cluster_desc_" + std::to_string(getpid()) + ".textproto";
```
Not a fan of this. We should just expose a Cabling Generator ctor that accepts a vector of hostnames and a `ClusterDescriptor` proto.

The other reason this is unclean is that the caller of this API already has a handle to the `cabling_descriptor_path` - so this means that we're just creating a temp copy of that file.

tt-asaigal commented on tools/scaleout/generate_mgd/generate_mgd.cpp:
```diff
@@ -0,0 +1,488 @@
+
+    // Create a single mesh descriptor that will be reused for all instances
+    auto* mesh_desc = mgd.add_mesh_descriptors();
+    mesh_desc->set_name("M0");  // Use M0 like in the example
```
M0 is a misnomer (Mesh 0). Would rather just declare it as `Mesh`.

tt-asaigal commented on tests/tt_metal/tt_fabric/fabric_router/test_cabling_descriptor_mgd_generation.cpp:
```diff
@@ -0,0 +1,249 @@
+
+    for (const auto conn_id : connections) {
+        const auto& conn = desc.get_connection(conn_id);
+        if (conn.nodes.size() < 2) {
```
SHouldn't this assert?

tt-asaigal commented on tools/scaleout/CMakeLists.txt:
```diff
@@ -126,23 +127,45 @@ target_link_libraries(
         cxxopts::cxxopts
 )
 
-add_executable(generate_cluster_descriptor src/generate_cluster_descriptor.cpp)
```
We shouldn't be removing the build for the cluster descriptor generation tool. I see a fair number of changes deleting this build step in this file.

tt-asaigal commented on tools/scaleout/generate_mgd/README.md:
```diff
@@ -0,0 +1,74 @@
+# Generate MGD Tool
+
+This tool automatically generates Mesh Graph Descriptors (MGDs) from cabling descriptors for fabric testing.
```
Please mention that we currently only support creating MGDs for multi-mesh topologies, based on the cabling descriptor. Each physical host corresponds to a mesh in the graph.

tt-asaigal reviewed (commented):
Core changes look good. Left a few comments for cleaning up the impl, build flow and docs.

Next steps for Auto MGD generation we can look into would include exposing more knobs for a user to play with:
- Relaxed vs Strict connectivity
- Generating an MGD for a sub-graph (specific host ids in the cabling descriptor)

jpanasiukTT commented on tools/scaleout/CMakeLists.txt:
```diff
@@ -75,6 +75,7 @@ target_link_libraries(
 target_link_libraries(
     run_cluster_validation
     PRIVATE
+        scaleout_tools
```
oh yeah dependency isn't strictly required, removed it

jpanasiukTT committed:
Refactor CablingGenerator with protobuf constructor

- Add constructor accepting ClusterDescriptor protobuf directly
- Extract common initialization logic into initialize_cluster helper
- Add forward declarations for proto types
- Enables MGD generation without temporary file I/O

jpanasiukTT committed:
Add generate_mgd tool for MGD generation from cabling descriptors

- Implement generate_mgd library with file and protobuf APIs
- Add CLI tool with cxxopts argument parsing
- Support multi-mesh topology generation
- Add comprehensive README with usage examples

jpanasiukTT committed:
Add MGD generation tests and CI integration

- Add unit tests for MGD generation from cabling descriptors
- Create generate_mgd_lib OBJECT library in CMake
- Integrate tests into CI via run_tools_tests.sh
- Add test configs for 16-node closetbox and dual T3K topologies
